### PR TITLE
[WIP] Initial ARM RyuJIT bring up

### PR DESCRIPTION
**Pull request by benpye**
This PR is to track my initial attempt to get the ARM RyuJIT backend to the point where it will execute a simple "Hello World". Currently this is to get feedback on the changes I am making to ensure I am not going in wrong direction or am doing anything obviously wrong to anyone who works on the JIT.

Currently when running a simple "Hello World" the JIT will successfully generate the `System.AppDomain:SetupDomain` method, begin to generate `System.AppDomain:SetupFusionStore` and then fail on an assertion, the full output of the debug coreclr with `COMPlus_JitDisasm=*` and COMPlus_DumpJittedMethods=1` is at https://gist.github.com/benpye/a4fcc0154c6d69efe875

### CODE REVIEW

File: src/jit/codegenarm.cpp [hunk 1]
```diff
@@ -152,7 +152,7 @@ void                CodeGen::genSetRegToIcon(regNumber     reg,
     else
     {
         // TODO-CQ: needs all the optimized cases
-        getEmitter()->emitIns_R_I(INS_mov, emitActualTypeSize(type), reg, val);
+        instGen_Set_Reg_To_Imm(emitActualTypeSize(type), reg, val);
```
Comment (BruceForstall) [inline src/jit/codegenarm.cpp]: This line subsumes the (val == 0) case also. In general, look at codegenarm64.cpp, and do whatever it has done. In this case, it only calls instGen_Set_Reg_To_Imm() for this function.

File: src/jit/codegenarm.cpp [hunk 2]
```diff
@@ -1012,7 +1022,19 @@ void                CodeGen::instGen_Set_Reg_To_Imm(emitAttr    size,
         else
 #endif // _TARGET_AMD64_
         {
-            getEmitter()->emitIns_R_I(INS_mov, size, reg, imm);
+            if ((imm & 0x0000ffff) == imm)
+            {
+                getEmitter()->emitIns_R_I(INS_mov, size, reg, imm);
+            }
+            else if ((flags != INS_FLAGS_SET) && ((imm & 0xffffffff) == imm))
```
Comment (BruceForstall) [inline src/jit/codegenarm.cpp]: ssize_t will be 32-bits on ARM, so this imm expression will always be true.

File: src/jit/codegenarm.cpp [hunk 3]
```diff
@@ -1498,7 +1520,76 @@ CodeGen::genCodeForTreeNode(GenTreePtr treeNode)
...
+                }
+                if (data->isContained() && !data->OperIsLeaf())
+                {
+                    NYI("RMW?");
+                }
+                else
+                {
+                    emit->emitInsMov(ins_Store(targetType), emitTypeSize(treeNode), treeNode);
+                }
+            }
```
Comment (BruceForstall) [inline src/jit/codegenarm.cpp]: It looks like you copied/modified this block from the AMD64 implementation. You might want to copy/modify from the ARM64 implementation instead. E.g., there's no read-modify-write on ARM.

**Comment by BruceForstall** [comment]
In general, the approach looks good.

Some general guidance:

When building up RyuJIT/ARM32, you have several places from which to grab code or ideas:
1. RyuJIT/x64: this is the fully functioning, full RyuJIT architecture compiler for x64. Useful to see what really works. Downside: uses x64 architecture, which has very different code from ARM32. Also: has 64-bit registers.
2. RyuJIT/ARM64: this is somewhat far along, but has a long way to go. Plus: it's more similar to ARM32 than x64. Downside: ARM64 is not ARM32. Plus, it's 64-bit.
3. legacy/ARM32: by "legacy", I mean the old code generator and register allocator. This is useful to see the exact code sequences that should be generated. But not too useful when figuring out how to generate them, in the new RyuJIT back-end architecture. However, it is 32-bit.

Picking and choosing code and ideas from among these is the way to go.

The first thing I would do is go through codegenarm.cpp and "clean it up". This was from an early experiment creating a RyuJIT architecture ARM32 back-end. A lot has changed since it was created, and it wasn't kept up-to-date. Pick up changes from the ARM64 or x64 back-ends, and clear out the cruft and add NYI. Then, start adding real code for functions as you need them.

File: src/jit/codegenarm.cpp (hunk 3 above)
Comment (benpye) [inline src/jit/codegenarm.cpp]: This is actually from the ARM64 implementation, I suppose that's why the NYI was "RMW?"

**Comment by benpye** [comment]
Thanks for the comments, after going through the methods present in codegenarm.cpp I have gotten past that assert which is good to see, however I feel I am missing something when I use R12 as a temporary register. In the generated code for SetupDomain it appears to be using R12 as though it was a non volatile register, as it sets R12, then uses it's value after a call ( https://gist.github.com/benpye/f042a230742a5c59b664#file-gistfile1-txt-L88 ), I assume I am missing something to indicate that R12 should be avoided in code gen?

**Comment by benpye** [comment]
Getting a little further now, https://gist.github.com/benpye/91406e28f48836e99a65 . The segfault seems to be due to the first argument to `System.String:Compare` being invalid, though I am struggling to pin down why as it's difficult to inspect the managed frames in gdb. https://gist.github.com/benpye/4e9b8c5e2d70173738bc is the full method for `System.AppDomain:PrepareDataForSetup` .

**Comment by benpye** [comment]
c430b00 gets execution to the first FCALL in this case System.Type:get_UnderlyingSystemType . This seems to be broken as the ARM vm assumes it can use R12 for PreStub related methods, so I think I'll have to change my usage of R12 in the JIT. It seems like it's going to be difficult to reserve the register in the lowering stage as it's needed in all cases where genHelperCall is used for example, which is called in a lot of cases in codegencommon.cpp . My issue then lies with the fact that the only other callee trash registers are R0-R3 which are used for arguments, any suggestions?

**Comment by benpye** [comment]
Found the issue more or less by accident, call lowering wasn't enabled for ARM so the base method was being called on `System.Type` and not the derived method on `System.RuntimeType`. I believe I am going to have difficulty getting further now as I am hitting the first case of an operation on a long, which RyuJIT does not have any support for under 32 bit, @BruceForstall you said you had a plan for this on x86 correctly? For reference, this is the calls generated currently:
```
Compiling    0 System.AppDomain::SetupDomain, IL size = 58, hsh=0x7a83b94f
Compiling    1 System.AppDomain::SetupFusionStore, IL size = 71, hsh=0xc5141dd8
Compiling    2 System.AppDomainSetup::VerifyDir, IL size = 31, hsh=0x12d702c7
Compiling    3 System.AppDomainSetup::SetupDefaults, IL size = 108, hsh=0x2a7c5db0
Compiling    4 System.String::LastIndexOfAny, IL size = 22, hsh=0xa01e2e37
Compiling    5 System.String::Substring, IL size = 144, hsh=0x85a5f80d
Compiling    6 System.String::InternalSubString, IL size = 102, hsh=0xc6f3096a
Compiling    7 System.String::Concat, IL size = 71, hsh=0x32fded48
Compiling    8 System.String::FillStringChecked, IL size = 63, hsh=0xfc9beb61
Compiling    9 System.AppDomain::PrepareDataForSetup, IL size = 189, hsh=0x5305b785
Compiling   10 System.AppDomainSetup::.ctor, IL size = 243, hsh=0x7785ad4b
Compiling   11 System.Collections.Generic.List`1[__Canon][System.__Canon]::.cctor, IL size = 12, hsh=0xce811fe1
Compiling   12 System.String::Compare, IL size = 253, hsh=0x493872fb
Compiling   13 System.String::CompareOrdinalIgnoreCaseHelper, IL size = 147, hsh=0x67dae402
Compiling   14 System.Collections.Generic.List`1[__Canon][System.__Canon]::Add, IL size = 74, hsh=0x2c52c7
Compiling   15 System.Collections.Generic.List`1[__Canon][System.__Canon]::EnsureCapacity, IL size = 62, hsh=0xeea03788
Compiling   16 System.Collections.Generic.List`1[__Canon][System.__Canon]::set_Capacity, IL size = 89, hsh=0x2ec3254f
Compiling   17 System.AppDomainSetup::SetCompatibilitySwitches, IL size = 71, hsh=0x3762fb27
Compiling   18 System.Collections.Generic.Dictionary`2[__Canon,__Canon][System.__Canon,System.__Canon]::.ctor, IL size = 45, hsh=0xe331a3f9
Compiling   19 System.Collections.Generic.EqualityComparer`1[__Canon][System.__Canon]::get_Default, IL size = 27, hsh=0x139ea37b
Compiling   20 System.Collections.Generic.EqualityComparer`1[__Canon][System.__Canon]::CreateComparer, IL size = 364, hsh=0xf2b12c08
Compiling   21 System.RuntimeType::IsAssignableFrom, IL size = 115, hsh=0x8064630a
Compiling   22 System.RuntimeType::get_UnderlyingSystemType, IL size = 2, hsh=0xab020624
Compiling   23 System.Collections.Generic.GenericEqualityComparer`1[__Canon][System.__Canon]::.ctor, IL size = 7, hsh=0x242f06ec
Compiling   24 System.Collections.Generic.List`1[__Canon][System.__Canon]::System.Collections.Generic.IEnumerable<T>.GetEnumerator, IL size = 12, hsh=0xec643fbc
Compiling   25 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   26 Enumerator[__Canon][System.__Canon]::MoveNext, IL size = 81, hsh=0xb1aa5dd0
Compiling   27 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   28 Enumerator[__Canon][System.__Canon]::get_Current, IL size = 7, hsh=0x45024f06
Compiling   29 System.Collections.Generic.Dictionary`2[__Canon,__Canon][System.__Canon,System.__Canon]::Insert, IL size = 398, hsh=0x70841fca
Compiling   30 System.Collections.Generic.Dictionary`2[__Canon,__Canon][System.__Canon,System.__Canon]::Initialize, IL size = 67, hsh=0x375de4b5
Compiling   31 System.Collections.HashHelpers::GetPrime, IL size = 90, hsh=0xcd87ecf0
Compiling   32 System.Collections.HashHelpers::.cctor, IL size = 24, hsh=0x58637e62
Compiling   33 System.Collections.Generic.GenericEqualityComparer`1[__Canon][System.__Canon]::GetHashCode, IL size = 24, hsh=0x665772c1
Compiling   34 System.String::GetHashCode, IL size = 169, hsh=0x5e996735
Compiling   35 Enumerator[__Canon][System.__Canon]::MoveNextRare, IL size = 59, hsh=0x366d774
Compiling   36 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   37 Enumerator[__Canon][System.__Canon]::Dispose, IL size = 1, hsh=0x7ce38031
Compiling   38 System.AppDomain::Setup, IL size = 1506, hsh=0x7d842baf
Compiling   39 System.Threading.Thread::GetDomain, IL size = 17, hsh=0x87066a31
Compiling   40 System.AppDomainSetup::GetConfigurationBytes, IL size = 27, hsh=0x4f67aab0
Compiling   41 KeyCollection[__Canon,__Canon][System.__Canon,System.__Canon]::.ctor, IL size = 23, hsh=0x5977702d
Compiling   42 KeyCollection[__Canon,__Canon][System.__Canon,System.__Canon]::System.Collections.Generic.IEnumerable<TKey>.GetEnumerator, IL size = 17, hsh=0xc3c8a64
Compiling   43 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   44 Enumerator[__Canon,__Canon][System.__Canon,System.__Canon]::MoveNext, IL size = 173, hsh=0x77da9652
Compiling   45 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   46 Enumerator[__Canon,__Canon][System.__Canon,System.__Canon]::get_Current, IL size = 7, hsh=0x7144ae44
Compiling   47 DomainNeutralILStubClass::IL_STUB_UnboxingStub, IL size = 30, hsh=0x4940e9a7
Compiling   48 Enumerator[__Canon,__Canon][System.__Canon,System.__Canon]::Dispose, IL size = 1, hsh=0x9e5b38d3
Compiling   49 System.String::Equals, IL size = 38, hsh=0x18ce0103
Compiling   50 System.String::EqualsHelper, IL size = 184, hsh=0x315cae05
Compiling   51 System.Text.StringBuilder::.ctor, IL size = 237, hsh=0xad28ce0a
Compiling   52 System.Text.StringBuilder::ThreadSafeCopy, IL size = 61, hsh=0xdd377bd4
Compiling   53 System.IO.Path::.cctor, IL size = 153, hsh=0x47b88b
Compiling   54 System.String::SplitInternal, IL size = 165, hsh=0xd9dcb289
Compiling   55 System.String::MakeSeparatorList, IL size = 208, hsh=0x2af61de1
Compiling   56 System.String::InternalSplitKeepEmptyEntries, IL size = 130, hsh=0xd8637935
Compiling   57 System.IO.Path::IsRelative, IL size = 41, hsh=0x792976e2
Compiling   58 System.IO.Path::NormalizePath, IL size = 1147, hsh=0x5a41ec4

Assert failure(PID 8402 [0x000020d2], Thread: 8402 [0x20d2]): Assertion failed 'NYI: "Arithmetic binary operators on TYP_LONG"' in 'System.IO.Path:NormalizePath(ref,bool,int,bool):ref' (IL size 1147)

    File: /home/ubuntu/git/coreclr/src/jit/lower.cpp Line: 404
    Image: /home/ubuntu/git/coreclr/bin/Product/Linux.arm.Debug/corerun


Program received signal SIGTRAP, Trace/breakpoint trap.
0xb6387e0c in DBG_DebugBreak () at /home/ubuntu/git/coreclr/src/pal/src/thread/context.cpp:1293
1293	    __asm__ __volatile__(".inst.w 0xde01");
(gdb) 
```

**Comment by mikedn** [comment]
> now as I am hitting the first case of an operation on a long, which RyuJIT does not have any support for under 32 bit, @BruceForstall you said you had a plan for this on x86 correctly? 

Note that there is some support for long operations, see `Lowering::DecomposeNode`. The question is if that's the current plan or just an old attempt.

**Comment by benpye** [comment]
@mikedn I'm assuming old attempt, the existing 32 bit stuff appears to either originate from the old ARM port (which has long since regressed, the code structure is vastly different now) or maybe from an x86 port, but at least since coreclr has been opened there hasn't been any progress on this front.

**Comment by BruceForstall** [comment]
Yes, there is a plan to implement long support in LSRA (and elsewhere in the RyuJIT back-end) for x86, and that would be usable by ARM as well. @CarolEidt has done some thinking and implementation on this, but its priority has been trumped by other things for a while now.

One thing you could consider to make progress is the following. How far does the JIT get with LEGACY_BACKEND defined? Presumably pretty far. So, build the LEGACY_BACKEND as the "main" JIT, then build the RyuJIT/ARM back-end as an "altjit" (build it with ALT_JIT defined). Call that protojit.dll. Then, run your tests with COMPLUS_AltJit=* (star means use altjit for all methods) and COMPLUS_AltJitName=protojit.dll (protojit.dll is actually the default). Then, when you hit an NYI assert, it will automatically fall back to the "main" JIT.

Another option is to build the VM with the interpreter enabled (define FEATURE_INTERPRETER), and use the interpreter as a fall-back. This is what is done on ARM64. I don't know how well the interpreter works on ARM32. Use it via:

set COMPLUS_INTERPRETERFALLBACK=1

(It's been a long time since I've tried this, so I'm not sure I have it 100% correct.)

**Comment by benpye** [comment]
Enabling the interpreter seemed to be fairly easy so that's what I've done for now. Get's me to the next issue so I guess I'll carry on from there.

**Comment by benpye** [comment]
Had some other stuff happen since I'd last looked at this, but having just rebased changes the state appears to be the same. When using the interpreter to get past the first long issue, I hit an assertion on the first PInvoke.

```
Compiling   76 DomainNeutralILStubClass::IL_STUB_PInvoke, IL size = 91, hsh=0x2f6692f9

Assert failure(PID 23409 [0x00005b71], Thread: 23409 [0x5b71]): Assertion failed 'varDsc->lvIsInReg() || varDsc->lvOnFrame || varDsc->lvRefCnt == 0' in 'DomainNeutralILStubClass:IL_STUB_PInvoke(struct,int):int' (IL size 91)

    File: /home/ubuntu/git/coreclr/src/jit/regalloc.cpp Line: 6844
    Image: /home/ubuntu/git/coreclr/bin/Product/Linux.arm.Debug/corerun


Program received signal SIGTRAP, Trace/breakpoint trap.
0xb636ae8c in DBG_DebugBreak ()
    at /home/ubuntu/git/coreclr/src/pal/src/thread/context.cpp:1291
1291        __asm__ __volatile__(".inst.w 0xde01");
```

**Comment by hqueue** [comment]
I'm also interested in ARM RyuJIT backend. Is there any updates recently?

**Comment by manu-silicon** [comment]
Have a look at #3977 for Linux ARM. This is not using fully Ryujit but it shows how much work has to be done.

**Comment by Petermarcu** [comment]
@benpye , what are the next steps on this one? We are trying not to just have PR's open for long periods of time. Is this on track to check in soon? If not, can we close it and track the conversation in an issue?

**Comment by benpye** [comment]
Probably best to close this one, it'll be way out of sync with the current
tree at this point, and it seems that the current plan for ARM is to use
the RyuJIT front end with the legacy backend anyway.

**Comment by Petermarcu** [comment]
@benpye , Thanks!

**Comment by BruceForstall** [comment]
@benpye Note that the JIT team would still like to end up with a RyuJIT port to ARM, which would allow us to delete the LEGACY_BACKEND code (and thus greatly simplify the JIT code). However, that might not happen (at Microsoft anyway) for a while. So in the meantime, some people are using the ARM LEGACY_BACKEND path.

**Pull request closed without merging**